Add dynamic params to attestation

justin1121 opened the pull request:
Can't disable verifying certificate chain right now and fake time isn't implemented yet. Will do a follow up PR to implement the fake time functionality.

justin1121 committed:
Add dynamic params to attestation

Can't disable verifying certificate chain right now and fake time isn't
implemented yet. Will do a follow up PR to implement the fake time
functionality.

kitschysynq commented on attest/attest.go:
```diff
@@ -20,6 +20,28 @@ import (
 	"github.com/veraison/go-cose"
 )
 
+type Params struct {
```
If `Params` and all of its members are exported, what's the purpose of the functional options?

justin1121 replied:
Could make them private. Wasn't sure how this pattern usually looks.

kitschysynq replied:
It's a good pattern! Typically you'd use it on a constructor, so maybe something like [this](https://go.dev/play/p/MN-orWMzGzq). One of the advantages of operating directly on a struct that does the work is that you have one less layer of plumbing you have to maintain when adding or removing functionality.

justin1121 replied:
Nice something like that makes sense. 

In the example, were you thinking Verifier struct would do all the attestation work or just the verifying certificate work? I think it makes sense if one struct was doing all the work because it would make it easier for the consumer to configure.

justin1121 replied:
Updated now let me know what you think.

justin1121 committed:
make params and its member private

justin1121 committed:
Implemented suggestion to make Verifier struct

Can pass custom root cert and current time. Setting current time can be used to
make verify certificate chain pass with an expired cert.

kitschysynq commented on attest/attest.go:
```diff
@@ -132,11 +161,18 @@ func Attest(attestation []byte, nonce []byte, rootCert *x509.Certificate) (*Atte
-			log.Errorf("Error verifying certificate chain: %v", err)
+	rootCert := v.rootCert
+	if rootCert == nil {
+		c, err := GetRootAWSCert()
```
Not a blocker: Since this involves a network access, it's worth considering whether we could effectively cache this cert by also storing it in `v.rootCert`.

justin1121 replied:
Good idea, just going to merge for right now. Will open an issue.
